datafusion/core/src/physical_plan/filter.rs

```diff
@@ -172,21 +172,41 @@ impl ExecutionPlan for FilterExec {
     /// predicate's selectivity value can be determined for the incoming data.
     fn statistics(&self) -> Statistics {
         let input_stats = self.input.statistics();
-        let analysis_ctx =
+        let mut analysis_ctx =
             AnalysisContext::from_statistics(self.input.schema().as_ref(), &input_stats);
 
         let predicate_selectivity = self
             .predicate
-            .boundaries(&analysis_ctx)
+            .boundaries(&mut analysis_ctx)
             .and_then(|bounds| bounds.selectivity);
```

Comment:
What if this looked something like:


```rust
 let predicate_selectivity = self
            .predicate
            .boundaries(analysis_ctx)
            .and_then(|analysis_ctx| analysis_ctx.bounds().selectivity);
```

Namely that `boundaries` took in a context and returned a potentially modified one? 

The reason I really worry about the `mut analysis_ctx` thing is when there is an operator that removes the intermediate column boundaries (like maybe `CASE` or some user defined function call) each place will have to remember to clear any intermediate boundaries built up in the context. 

The idea of tracking intermediate column boundaries looks very cool

Reply:
> Namely that boundaries took in a context and returned a potentially modified one?

Thinking about it, that sounds really intriguing. Will give it a try, thank you for your suggestion and feedback on this @alamb!